dotcom-rendering/src/web/components/SlotBodyEnd.importable.tsx

```diff
@@ -88,12 +93,15 @@ export const SlotBodyEnd = ({
 	isPaidContent,
 	tags,
 	contributionsServiceUrl,
-	brazeMessages,
 	idApiUrl,
 	stage,
 	pageId,
 	keywordsId,
-}: Props) => {
+}: Omit<Props, 'isDev' | 'switches' | 'isSensitive'>) => {
```

Comment:
We did try to avoid it, but you then need two `Props` types which are nearly identical. Either that or you have to use `PropsA & PropsB`, which I wouldn't say is any clearer.

Any idea on how we can tackle this?

Reply:
To be honest, I can't suggest an alternative because I (genuinely) don't understand the code or what problem it is trying to solve. My guess is though that if DCR has been able to manage this long without this feature there probably is a way to make it work

Reply:
@mxdvl on intersection types, we could add clarity and maybe have something to share elsewhere.

e.g.
```TS
// AB.ts
export type Props = {
  switches: Switches;
  isSensitive: boolean;
  isDev: boolean;
}

// SlotBodyEnd.tsx
import { ABProps } from './AB.ts';
type Props = {
  contentType: string;
  // ...
}

const SlotBodyEndWithAB = (props: Props & ABProps) = {}
```

Reply:
@oliverlloyd in this case I think we’re coming up against some of the limitations of the Island pattern. The reason DCR managed this long without this feature is because we could confidently use composition. We found ourselves more comfortable with a single child to an Island, opting for a pattern where `SlotBodyEnd` contains its own `ABProvider`:

```tsx
// Layout.tsx
<Island clientOnly={true}>
  <SlotBodyEnd
    propA="A"
    propB="B"
    propC="C"
    propD="D"
    /* … */
    propM="M"
    propN="N"
  />
</Island>

// SlotBodyEnd.tsx
export const SlotBodyEnd = ({ /* … */ }: Props) => {
	return (
		<WithABProvider
			propA={A}
			propF={F}
			propN={N}
		>
			<SlotBodyEndWithAB
				propA="A"
				propB="B"
				propC="C"
				propD="D"
				/* … */
				propM="M"
			/>
		</WithABProvider>
	);
};
```

The alternative to this would be to allow ABProvider to be importable, but that seems odd.

Reply:
Is the problem here that you don't want to have two sets of props for two different components? And want to connect the different types together and create one from the other? Asking because I am still a bit on the edge of understanding the goal with `Omit`.

If my assumption here is right then I can see the benefit. You don't have to create a new type and have instead got a sort of abstraction that creates it for you. Another benefit is you're coupling the types together so any changes to the root type will be carried down into the child version of it.

The counter to that is we're adding complexity which we could avoid by having duplication and that the benefit of coupling is also a constraint.

> limitations of the Island pattern

I don't think this is entirely true. The reason we're using composition is not really related to the islands pattern but is instead caused by the constraint that hooks cannot be called conditionally. The islands pattern really only means we need to add an `ABProvider`, composition is how we would need to add that no matter if we were using an island or not.

In general though, I am happy to adopt these sorts of features if there were team consensus around them. You can have these sorts of discussions endlessly and there's really no right or wrong here.

Reply:
> If my assumption here is right then I can see the benefit. You don't have to create a new type and have instead got a sort of abstraction that creates it for you. Another benefit is you're coupling the types together so any changes to the root type will be carried down into the child version of it.
>
> The counter to that is we're adding complexity which we could avoid by having duplication and that the benefit of coupling is also a constraint.

By placing the ABProvider inside this component, we’re fully coupled, so the types might as well reflect it. There might be something I’m not seeing here. I’ll have a go at making `WithABProvider` an importable.

Reply:
I think some of the complexity lies in how we have decided Islands should be implemented if the need access to the `useAB` hook.

For components to work, with AB hooks, in islands, we need to re-include the `WithABProvider`. We have chosen to make this the responsibility of the component rather than the island (can't find the documentation for this).  

This means we have the pattern:
```TS
BootReact
├─ 
├─ WithABProvider
│  ├─ Island
│  │  ├─ ComponentWithABProvider
│  │  │  ├─ WithABProvider
│  │  │  │  ├─ Component
```

That means `ComponentWithABProvider` require both the types for `WithABProvider` and the `Component`.

To this end, `ComponentWithABProvider` would have the type of `WithABProvider & ComponentProps`. This to me is quite easy to understand, or by using the Omit pattern, although for clarity I might use `Omit<Props, ABProps>`.

The parent component would always have to implement the `<Island><ComponentWithABProvider>` if the component uses the `useAB` hook.

Maybe there is something we can do to simplify this?